lib/web/inventory_test.go

```diff
@@ -0,0 +1,244 @@
+				Hostname: "host-5",
+				Version:  "18.5.0",
+				Services: []types.SystemRole{types.RoleProxy},
+			},
+		},
+		{
+			ResourceHeader: types.ResourceHeader{
+				Metadata: types.Metadata{Name: "instance-6"},
+			},
+			Spec: types.InstanceSpecV1{
+				Hostname: "host-6",
+				Version:  "18.6.0",
+				Services: []types.SystemRole{types.RoleAuth},
+			},
+		},
+	}
+	for _, instance := range instances {
+		require.NoError(t, env.server.Auth().UpsertInstance(ctx, instance))
+	}
+
+	// Create mock bot instances
+	bots := []*machineidv1.BotInstance{
+		{
+			Metadata: &headerv1.Metadata{Name: "bot-1"},
+			Spec: &machineidv1.BotInstanceSpec{
+				BotName:    "bot-1",
+				InstanceId: "bot-1",
+			},
+			Status: &machineidv1.BotInstanceStatus{
+				LatestHeartbeats: []*machineidv1.BotInstanceStatusHeartbeat{
+					{
+						RecordedAt: timestamppb.Now(),
+						Version:    "18.7.0",
+					},
+				},
+			},
+		},
+		{
+			Metadata: &headerv1.Metadata{Name: "bot-2"},
+			Spec: &machineidv1.BotInstanceSpec{
+				BotName:    "bot-2",
+				InstanceId: "bot-2",
+			},
+			Status: &machineidv1.BotInstanceStatus{
+				LatestHeartbeats: []*machineidv1.BotInstanceStatusHeartbeat{
+					{
+						RecordedAt: timestamppb.Now(),
+						Version:    "18.8.0",
+					},
+				},
+			},
+		},
+		{
+			Metadata: &headerv1.Metadata{Name: "bot-3"},
+			Spec: &machineidv1.BotInstanceSpec{
+				BotName:    "bot-3",
+				InstanceId: "bot-3",
+			},
+			Status: &machineidv1.BotInstanceStatus{
+				LatestHeartbeats: []*machineidv1.BotInstanceStatusHeartbeat{
+					{
+						RecordedAt: timestamppb.Now(),
+						Version:    "18.9.0",
+					},
+				},
+			},
+		},
+	}
+	for _, bot := range bots {
+		_, err := env.server.Auth().BotInstance.CreateBotInstance(ctx, bot)
+		require.NoError(t, err)
+	}
+
+	// Create user with required permissions
+	username := "test-user"
+	role, err := types.NewRole(services.RoleNameForUser(username), types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			Rules: []types.Rule{
+				types.NewRule(types.KindInstance, []string{types.VerbRead, types.VerbList}),
+				types.NewRule(types.KindBotInstance, []string{types.VerbRead, types.VerbList}),
+			},
+		},
+	})
+	require.NoError(t, err)
+	pack := env.proxies[0].authPack(t, username, []types.Role{role})
+
+	endpoint := pack.clt.Endpoint("webapi", "sites", clusterName, "instances")
+
+	// Wait for the inventory cache to be healthy before making requests
+	require.Eventually(t, func() bool {
+		invCache := env.server.Auth().GetInventoryCache()
+		return invCache != nil && invCache.IsHealthy()
+	}, 10*time.Second, 100*time.Millisecond, "inventory cache failed to become healthy")
+
+	// Test listing with no params
+	resp, err := pack.clt.Get(ctx, endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var listResp listUnifiedInstancesResponse
+	require.NoError(t, json.Unmarshal(resp.Bytes(), &listResp))
+	require.Len(t, listResp.Instances, 9, "should have 9 items (6 instances + 3 bots)")
```

Comment:
Is inventory cache health alone enough to guarantee us that all the instances have been populated? Should we instead place the initial request in an eventually to ensure that we see all the expected items before continuing? 
```suggestion
	// The initial requests are in the eventually to ensure that the inventory
	// cache is healthy, has processed all put events, and that all the instances
	// are present in the cache.
	require.EventuallyWithT(t, func(t *assert.CollectT) {
			// Test listing with no params
			resp, err := pack.clt.Get(ctx, endpoint, url.Values{})
			require.NoError(t, err)

			var listResp listUnifiedInstancesResponse
			require.NoError(t, json.Unmarshal(resp.Bytes(), &listResp))
			require.Len(t, listResp.Instances, 9, "should have 9 items (6 instances + 3 
	}, 10*time.Second, 100*time.Millisecond, "inventory cache failed to become healthy")

bots)")
```